src/main/kotlin/com/wgslfuzz/core/Resolver.kt

```diff
@@ -124,42 +122,46 @@ sealed interface ResolvedEnvironment {
 }
 
 private class ScopeImpl(
-    override val parent: ScopeImpl?,
+    private val previous: ScopeImpl? = null,
+    // This is from the top global level of scope. 0 is the top level. 1 is one level below and so on
+    private val level: Int = 0,
+    private val scopeEntry: Pair<String, ScopeEntry>? = null,
```

Comment:
A comment on why this can be null would help. I think it's because when entering a new construct we need a "dud" scope that simply points to the previous scope, and we may not yet have any declaration, is that right?